src/logger/InteractiveLogger.js

```diff
@@ -0,0 +1,263 @@
+      this.isCursorAboveTasksSection = true
+    }
+  }
+
+  clearTasks() {
+    this.tasks = []
+  }
+
+  /**
+   * @param {string[]} args
+   */
+  log(...args) {
+    this.moveCursorToAboveTasksSection()
+    this.tty.clearScreenDown()
+    this.tty.write(args.join(' ') + '\n')
+    this.printTasks()
+  }
+
+  /**
+   * @param {string[]} args
+   */
+  logErr(...args) {
+    this.log(...args)
+  }
+
+  /**
+   * @param {string} headline
+   * @param {{ error?: Error, detail?: string }} [more]
+   * @returns {never}
+   */
+  fail(headline, more) {
+    this.tty.write('\n\n')
+    this.tty.write(formatFailMessage(headline, more))
+    this.tty.write('\n')
+    process.exit(1)
+  }
+
+  /**
+   * @param {string[]} args
+   */
+  info(...args) {
+    this.log(formatInfoMessage(...args))
+  }
+
+  /**
+   * @param {string[]} args
+   */
+  note(...args) {
+    this.log(formatNoteMessage(...args))
+  }
+
+  /**
+   * @param {string} message
+   */
+  success(message) {
+    this.log(formatSuccessMessage(message))
+  }
+
+  /**
+   * @param {string} taskName
+   * @returns {import('../types.js').TaskLogger}
+   */
+  task(taskName) {
+    let start = Date.now()
+    const color = getColorForString(taskName)
+    const prefix = color.fg(`${taskName} `)
+
+    this.moveCursorToAboveTasksSection()
+
+    /** @type {Task} */
+    const task = {
+      prefix: taskName,
+      coloredPrefix: prefix,
+      lastLogLine: '',
+      status: 'waiting',
+      startedAt: start,
+    }
+
+    this.tasks.push(task)
+    this.printTasks()
+
+    let isDone = false
+    const assertNotDone = () => {
+      if (isDone) {
+        throw new Error('This task has already been completed')
+      }
+    }
+
+    /** @type {string[]} */
+    const bufferedLogLines = []
+
+    const log = (/** @type {TaskStatus} */ status, /** @type {string[]} */ ...args) => {
+      assertNotDone()
+
+      // if the status changes, we need to reprint the tasks immediately. if
+      // not, we can just wait for our interval loop to do it
+      const didStatusChange = task.status !== status
+      if (didStatusChange) {
+        this.moveCursorToAboveTasksSection()
+        task.status = status
+      }
+
+      const message = args.join(' ')
+      bufferedLogLines.push(message)
+      const lastLine = lastNonEmptyLineIfPossible(message)
+      if (lastLine) {
+        task.lastLogLine = lastLine
+      }
+
+      if (didStatusChange) {
+        this.printTasks()
+      }
+    }
+
+    const complete = (/** @type {TaskStatus} */ status, /** @type {string} */ message) => {
+      assertNotDone()
+      this.moveCursorToAboveTasksSection()
+      this.tty.clearScreenDown()
+      task.status = status
+      const lastLine = lastNonEmptyLineIfPossible(message)
+      if (lastLine) {
+        task.lastLogLine = lastLine
+      }
+
+      bufferedLogLines.push(message)
+      this.tty.write(color.fg(':: ') + color.bg().bold(` ${taskName} `) + color.fg(' ::') + '\n')
```

Comment:
This feels kinda visually intense. Might try tweaking it after this gets merged, but if you wanna give it a go first please do :)